src/toxcore/dht/server.rs

```diff
@@ -180,25 +190,59 @@ impl Server {
 
         Box::new(pings_stream.for_each(|()| Ok(())))
     }
-    /// send NodesRequest to random peer every 20 seconds
-    pub fn send_nodes_req(&self, friend_pk: PublicKey) -> IoFuture<()> {
-        let mut state = self.state.write();
-        if let Some(peer) = state.kbucket.get_random_node() {
-            let client = state.peers_cache.entry(peer.pk).or_insert_with(ClientData::new);
 
-            let payload = NodesRequestPayload {
-                pk: friend_pk,
-                id: client.new_ping_id(),
-            };
-            let nodes_req = DhtPacket::NodesRequest(NodesRequest::new(
-                &precompute(&peer.pk, &self.sk),
-                &self.pk,
-                payload
-            ));
-            self.send_to(peer.saddr, nodes_req)
-        } else {
-            Box::new(future::ok(()))
+    /// Send NodesRequest to random peer every 20 seconds
+    pub fn periodical_nodes_req(&self) -> IoFuture<()> {
+        let peer = match self.get_random_node() {
+            None => {
+                return Box::new( future::ok(()))
+            },
+            Some(peer) => peer,
+        };
+
+        self.send_nodes_req(peer, None)
+    }
+    
+    // Get random node from kbucket
+    fn get_random_node(&self) -> Option<PackedNode> {
+        let state = self.state.read();
+        state.kbucket.get_random_node()
+    }
+
+    // Send NodesRequest to peer
+    // sendback_node is the node that it will send back the response to (set to None to disable this)
+    fn send_nodes_req(&self, target_peer: PackedNode, sendback_node: Option<PackedNode>) -> IoFuture<()> {
+        // Check if packet is going to be sent to ourself.
+        if self.pk == target_peer.pk {
+            return Box::new(
+                future::err(
+                    Error::new(ErrorKind::Other, "friend's pk is mine")
+                )
+            )
         }
+
+        let mut state = self.state.write();
+        let _client = state.peers_cache.entry(target_peer.pk).or_insert_with(ClientData::new);
+
+        let data = match sendback_node {
+            None => vec![target_peer],
+            Some(peer) => vec![target_peer, peer],
+        };
+
+        let mut ping_checker = self.ping_checker.write();
+        let ping_id = ping_checker.add(data);
```

Comment:
That is not obvious why do you add `data` with 2 nodes while

```
// sendback_node is the node that it will send back the response to (set to None to disable this)
```

and why do I really care which node will send me it and why PingChecker should care about it